rclpy/src/rclpy/_rclpy_signal_handler.c

```diff
@@ -0,0 +1,288 @@
+static void
+call_original_signal_handler(int signo)
+{
+  if (NULL != g_original_signal_handler) {
+    g_original_signal_handler(signo);
+  }
+}
+
+/// Catch signals
+/**
+ * This triggers guard conditions when a signal is received.
+ * These wake executors currently blocked in `rcl_wait`.
+ */
+static void catch_function(int signo)
+{
+  if (NULL == g_guard_conditions || NULL == g_guard_conditions[0]) {
+    // There may have been another signal handler chaining to this one when we last tried to
+    // restore the old signal handler.
+    // Try to restore the old handler again.
+    restore_original_signal_handler();
+    call_original_signal_handler(signo);
+    return;
+  }
+
+  // Trigger python signal handlers
+  rcl_guard_condition_t ** pgc = g_guard_conditions;
+  while (NULL != *pgc) {
+    rcl_ret_t ret = rcl_trigger_guard_condition(*pgc);
+    if (ret != RCL_RET_OK) {
+      // TODO(sloretz) find signal safe way to tell the world an error occurred
+      rcl_reset_error();
+    }
+    ++pgc;
+  }
+
+  // Chain signal handlers
+  call_original_signal_handler(signo);
+}
+
+/// Register a guard condition to be triggered when SIGINT is received.
+/**
+ * On failure, an exception is raised and NULL is returned if:
+ *
+ * Raises ValueError if the argument is not a guard condition handle
+ * Raises ValueError if the argument was already registered
+ *
+ * \param[in] pygc a guard condition pycapsule
+ * \return None
+ */
+static PyObject *
+rclpy_register_sigint_guard_condition(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  // Expect a pycapsule with a guard condition
+  PyObject * pygc;
+  if (!PyArg_ParseTuple(args, "O", &pygc)) {
+    return NULL;
+  }
+
+  rcl_guard_condition_t * gc = (rcl_guard_condition_t *)PyCapsule_GetPointer(
+    pygc, "rcl_guard_condition_t");
+  if (!gc) {
+    return NULL;
+  }
+
+  // Figure out how big the list currently is
+  size_t count_gcs = 0;
+  if (NULL != g_guard_conditions) {
+    while (NULL != g_guard_conditions[count_gcs]) {
+      if (gc == g_guard_conditions[count_gcs]) {
+        PyErr_Format(PyExc_ValueError, "Guard condition was already registered");
+        return NULL;
+      }
+      ++count_gcs;
+    }
+  }
+
+  // Current size of guard condition list: count_gcs + 1 (sentinel value)
+  // Allocate space for one more guard condition: count_cs + 1 (new gc) + 1 (sentinel value)
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_guard_condition_t ** new_gcs =
+    allocator.allocate(sizeof(rcl_guard_condition_t *) * (count_gcs + 2), allocator.state);
+
+  // populate the new guard condition list, ending with a sentinel of NULL
+  for (size_t i = 0; i < count_gcs; ++i) {
+    new_gcs[i] = g_guard_conditions[i];
+  }
+  new_gcs[count_gcs] = gc;
+  new_gcs[count_gcs + 1] = NULL;
+
+  // Swap the lists and free the old
+  rcl_guard_condition_t ** old_gcs = g_guard_conditions;
+  // Assumes this assignment is atomic
+  g_guard_conditions = new_gcs;
```

Comment:
@sloretz hmm, I'm not sure this is always true. It depends a lot on how this is compiled for each of our supported processor architectures. And that's even before adding SMP and cache coherence to the equation. 

Why not using OS specific atomics (i.e. abstracted CPU specific atomics)?

Reply:
I didn't find an atomic for pointer types. Maybe I could use a different type that is at least as big as an `rcl_guard_condition **` and cast to it?

Reply:
Gah, I didn't realize any type could be made atomic. 35388745423aad221846dfc7e1d0195da9e3c0ae uses an atomic type and methods. IIUC the signal handler could deadlock if this type is not lock-free, so the code issues a python warning once if that's the case.

Reply:
Well, `gcc` has some [intrinsics](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) that `llvm` seems to support as well. As usual, `msvc` has it's [own stuff](https://docs.microsoft.com/en-us/windows/desktop/api/winnt/nf-winnt-interlockedexchangepointer). We should verify against CI though.

Reply:
Oh, I didn't know we had an `rcutils` helper, nice! Though in that case I think we should be using `rcutils_atomic_*` macros/functions instead of `atomic_*` ones.

Reply:
Using `rcutils_atomic_*` in a2b1aaf79663a74ce168fe3d270ad69d279b99a8